extensions/amp-ad-network-doubleclick-impl/0.1/doubleclick-a4a-config.js

```diff
@@ -77,55 +83,116 @@ export const BETA_ATTRIBUTE = 'data-use-beta-a4a-implementation';
 /** @const {string} */
 export const BETA_EXPERIMENT_ID = '2077831';
 
-/**
- * @param {!Window} win
- * @param {!Element} element
- * @returns {boolean}
- */
-export function doubleclickIsA4AEnabled(win, element) {
-  if ('useSameDomainRenderingUntilDeprecated' in element.dataset ||
-      element.hasAttribute('useSameDomainRenderingUntilDeprecated') ||
-      !isGoogleAdsA4AValidEnvironment(win)) {
-    return false;
+let singleton;
+
+export class DoubleclickA4aEligibility {
+  constructor() {
+    if (!singleton) {
+      singleton = this;
+    }
+
+    return singleton;
   }
-  if (element.hasAttribute(BETA_ATTRIBUTE)) {
-    addExperimentIdToElement(BETA_EXPERIMENT_ID, element);
-    dev().info(TAG, `beta forced a4a selection ${element}`);
-    return true;
+
+  supportsCrypto(win) {
+    return supportsNativeCrypto(win);
   }
-  // See if in holdback control/experiment.
-  let experimentId;
-  const urlExperimentId = extractUrlExperimentId(win, element);
-  if (urlExperimentId != undefined) {
-    experimentId = URL_EXPERIMENT_MAPPING[urlExperimentId];
-    dev().info(
-        TAG, `url experiment selection ${urlExperimentId}: ${experimentId}.`);
-  } else {
-    // Not set via url so randomly set.
+
+  isCdnProxy(win) {
+    const googleCdnProxyRegex =
+      /^https:\/\/([a-zA-Z0-9_-]+\.)?cdn\.ampproject\.org/;
+    return !!googleCdnProxyRegex.test(win.location.origin);
+  }
+
+  isA4aEnabled(win, element) {
+    element = element;
+    if ('useSameDomainRenderingUntilDeprecated' in element.dataset ||
+        element.hasAttribute('useSameDomainRenderingUntilDeprecated') ||
+        !this.supportsCrypto(win)) {
+      return false;
+    }
+    if (!this.isCdnProxy(win)) {
+      return this.selectExperiment(win, element, [
+        DOUBLECLICK_EXPERIMENT_FEATURE.CANONICAL_CONTROL,
+        DOUBLECLICK_EXPERIMENT_FEATURE.CANONICAL_EXPERIMENT,
+      ], [DOUBLECLICK_EXPERIMENT_FEATURE.CANONICAL_EXPERIMENT],
+          DFP_CANONICAL_FF_EXPERIMENT_NAME);
+    }
+    if (element.hasAttribute(BETA_ATTRIBUTE)) {
+      addExperimentIdToElement(BETA_EXPERIMENT_ID, element);
+      dev().info(TAG, `beta forced a4a selection ${element}`);
+      return true;
+    }
+    // See if in holdback control/experiment.
+    let experimentId;
+    const urlExperimentId = extractUrlExperimentId(win, element);
+    if (urlExperimentId != undefined) {
+      experimentId = URL_EXPERIMENT_MAPPING[urlExperimentId];
+      dev().info(
+          TAG, `url experiment selection ${urlExperimentId}: ${experimentId}.`);
+      return this.addAndForceExperiment(win, element, experimentId, [
```

Comment:
this function forces the experimentId given (if present) and then returns if the experimentId is in the array?  Wouldn't that cause holdback external control to use FF?